### PR TITLE
Feature/check logging feature flags memory

### PR DESCRIPTION
**Pull request by supercid**
## Checklist:
- [ ] My code follows the code style of this project.
- [ ] I have updated the documentation accordingly.
- [ ] All new and existing tests passed.
- [ ] I have assigned the correct milestone or created one if non-existent.
- [ ] I have correctly labeled this pull request.
- [ ] I have linked the corresponding issue in this description.
- [ ] I have updated the corresponding Jira ticket.
- [ ] I have requested a review from at least 2 reviewers
- [ ] I have checked the base branch of this pull request
- [ ] I have checked my code for any possible security vulnerabilities

### CODE REVIEW

File: Model/Service/Index.php [hunk 1]
```diff
@@ -180,14 +195,28 @@ public function updateOrCreateDirtyEntity(Product $product, Store $store)
...
         $lastPage = $collection->getLastPageNumber();
         $curPage = 1;
         $totalDirty = 0;
+        $updatesEnabled = $this->nostoDataHelper->isProductUpdatesEnabled();
+        $maxMemPercentage = $this->nostoDataHelper->getIndexerMemory();
+        if (!$updatesEnabled) {
+            $this->logger->info('Skipping product sync since product updates via API are disabled');
+        }
         do {
+            if (NostoMemUtil::getPercentageUsedMem() >= $maxMemPercentage) {
```
Comment (hannunosto) [inline Model/Service/Index.php]: Create a private method `checkMemoryConsumption()` and do the checks there.

File: Model/Service/Index.php [hunk 2]
```diff
@@ -234,6 +267,14 @@ public function handleProductSync(NostoIndexCollection $collection, Store $store
                 )
             );
             do {
+                if (NostoMemUtil::getPercentageUsedMem() >= $maxMemPercentage) {
```
Comment (hannunosto) [inline Model/Service/Index.php]: Move to `checkMemoryConsumption `

File: Model/Service/Index.php [hunk 3]
```diff
@@ -400,10 +446,19 @@ public function handleProductDeletion(NostoIndexCollection $collection, Store $s
         if ($account instanceof NostoSignupAccount === false) {
             throw new NostoException(sprintf('Store view %s does not have Nosto installed', $store->getName()));
         }
+        $maxMemPercentage = $this->nostoDataHelper->getIndexerMemory();
         $collection->setPageSize(self::PRODUCT_DELETION_BATCH_SIZE);
         $lastPage = $collection->getLastPageNumber();
         $curPage = 1;
         do {
+            if (NostoMemUtil::getPercentageUsedMem() >= $maxMemPercentage) {
```
Comment (hannunosto) [inline Model/Service/Index.php]: Move to `checkMemoryConsumption`

File: Model/Indexer/Data.php [hunk 4]
```diff
@@ -61,37 +64,48 @@ class Data implements IndexerActionInterface, MviewActionInterface
...
     {
         $storesWithNosto = $this->nostoHelperAccount->getStoresWithNosto();
         foreach ($storesWithNosto as $store) {
             $indexCollection = $this->getCollection($store);
-            $this->nostoServiceIndex->handleDirtyProducts($indexCollection, $store);
+            try {
+                $this->nostoServiceIndex->handleDirtyProducts($indexCollection, $store);
+            } catch (MemoryOutOfBoundsException $e) {
+                $this->nostoLogger->error($e->getMessage());
+                throw $e;
```
Comment (hannunosto) [inline Model/Indexer/Data.php]: How does the indexer behave if we throw the exception here? We don't want to terminate the other index processing. Maybe just log the exception and not throw anything.

File: Model/Indexer/Data.php [hunk 5]
```diff
@@ -109,14 +123,19 @@ public function executeRow($id)
...
     {
         $storesWithNosto = $this->nostoHelperAccount->getStoresWithNosto();
         foreach ($storesWithNosto as $store) {
             $collection = $this->getCollection($store, $ids);
-            $this->nostoServiceIndex->handleDirtyProducts($collection, $store);
+            try {
+                $this->nostoServiceIndex->handleDirtyProducts($collection, $store);
+            } catch (MemoryOutOfBoundsException $e) {
+                $this->nostoLogger->error($e->getMessage());
+                throw $e;
```
Comment (hannunosto) [inline Model/Indexer/Data.php]: https://github.com/Nosto/nosto-magento2/pull/525/files#r320109349

File: Model/Service/Index.php [hunk 6]
```diff
@@ -276,6 +304,8 @@ public function handleProductSync(NostoIndexCollection $collection, Store $store
                     $store->getCode()
                 )
             );
+        } catch (MemoryOutOfBoundsException $e) {
```
Comment (hannunosto) [inline Model/Service/Index.php]: This logic is redundant. If you want this exception to be thrown from this method, simply don't catch it.

File: Model/Service/Index.php [hunk 7]
```diff
@@ -264,6 +290,8 @@ public function handleProductSync(NostoIndexCollection $collection, Store $store
                 }
                 ++$currentPage;
             } while ($currentPage <= $pages);
+        } catch (MemoryOutOfBoundsException $e) {
+            throw $e;
```
Comment (hannunosto) [inline Model/Service/Index.php]: This logic is redundant. If you want this exception to be thrown from this method, simply don't catch it.

File: Model/Service/Index.php (hunk 6 above)
Comment (supercid) [reply Model/Service/Index.php]: We need to bubble up the exception in order to stop the indexing process. They must be handled in the `execute` method

**Pull request merged**